### PR TITLE
Fix Sphinx links in documentation

### PR DESCRIPTION
**Pull request by hramezani**
Fixes https://github.com/urllib3/urllib3/issues/2378
Closes #2416 

Most of the documentation issues are related to types, maybe we can ignore them in `nitpick_ignore`.
What do you think @pquentin and  @sethmlarson?

```
urllib3/contrib/socks.py:docstring of urllib3.contrib.socks.SOCKSHTTPSConnectionPool:: WARNING: py:class reference target not found: Literal
urllib3/connection.py:docstring of urllib3.connection.HTTPSConnection.set_cert:: WARNING: py:class reference target not found: Literal
urllib3/connectionpool.py:docstring of urllib3.connectionpool.HTTPConnectionPool.urlopen:: WARNING: py:class reference target not found: urllib3.util.request._TYPE_FAILEDTELL
urllib3/connectionpool.py:docstring of urllib3.connectionpool.HTTPSConnectionPool:: WARNING: py:class reference target not found: Literal
urllib3/exceptions.py:docstring of urllib3.exceptions.HeaderParsingError:: WARNING: py:class reference target not found: email.errors.MessageDefect
urllib3/poolmanager.py:docstring of urllib3.ProxyManager.pools:: WARNING: py:class reference target not found: urllib3._collections.RecentlyUsedContainer
urllib3/response.py:docstring of urllib3.response.BaseHTTPResponse.get_redirect_location:: WARNING: py:class reference target not found: Literal
urllib3/response.py:docstring of urllib3.response.BaseHTTPResponse.geturl:: WARNING: py:class reference target not found: Literal
urllib3/response.py:docstring of urllib3.response.BaseHTTPResponse.info:: WARNING: py:class reference target not found: urllib3._collections.HTTPHeaderDict
urllib3/response.py:docstring of urllib3.response.HTTPResponse:: WARNING: py:class reference target not found: http.client.HTTPMessage
urllib3/response.py:docstring of urllib3.response.BaseHTTPResponse.get_redirect_location:: WARNING: py:class reference target not found: Literal
urllib3/response.py:docstring of urllib3.response.BaseHTTPResponse.geturl:: WARNING: py:class reference target not found: Literal
urllib3/response.py:docstring of urllib3.response.BaseHTTPResponse.info:: WARNING: py:class reference target not found: urllib3._collections.HTTPHeaderDict
urllib3/util/retry.py:docstring of urllib3.util.retry.Retry:: WARNING: py:class reference target not found: urllib3.util.retry.RequestHistory
urllib3/util/retry.py:docstring of urllib3.util.retry.Retry:109: WARNING: py:attr reference target not found: Retry.backoff_max
urllib3/util/retry.py:docstring of urllib3.util.retry.Retry:124: WARNING: py:class reference target not found: RequestHistory
/home/docs/.pyenv/versions/3.7.9/lib/python3.7/ssl.py:docstring of ssl.SSLContext:1: WARNING: py:class reference target not found: _ssl._SSLContext
urllib3/util/__init__.py:docstring of urllib3.util.ssl_.resolve_cert_reqs:: WARNING: py:class reference target not found: VerifyMode
urllib3/util/__init__.py:docstring of urllib3.util.ssl_.ssl_wrap_socket:: WARNING: py:class reference target not found: SSLTransportType
```

### CODE REVIEW

**Comment by pquentin** [comment]
I think we can look into this case by case. For example, not having a link to `Literal` in `Literal[False]` is fine.

The main question I think is: do we want to show type annotation at all in docs? If we don't then this will fix issues such as `urllib3._collections.RecentlyUsedContainer` not being documented.

Capture d’écran 2022-05-06 à 21 28 07

I'm not sure if this is actually useful. Maybe try `autodoc_typehints = "description"` as mentioned in https://github.com/pytest-dev/pytest/issues/8405?

By the way, I was surprised to see `RequestMethods`  in `urllib3.request()`, I don't think this is the right place.

File: docs/reference/urllib3.connection.rst [hunk 1]
```diff
@@ -1,6 +1,10 @@
 Connections
 ===========
 
+.. autoclass:: urllib3.connection.ProxyConfig
```
Comment (sethmlarson) [inline docs/reference/urllib3.connection.rst:4]: Let's list `ProxyConfig` in its own subsection after HTTPConnection classes?

File: docs/reference/urllib3.poolmanager.rst [hunk 2]
```diff
@@ -5,6 +5,12 @@ Pool Manager
     :members:
     :undoc-members:
     :show-inheritance:
+    :inherited-members:
+
+.. autoclass:: urllib3.poolmanager.PoolKey
```
Comment (sethmlarson) [inline docs/reference/urllib3.poolmanager.rst:10]: Let's move `PoolKey` to the bottom of this page.

File: docs/reference/urllib3.request.rst [hunk 3]
```diff
@@ -1,4 +1,9 @@
 urllib3.request()
 =================
 
+.. autoclass:: urllib3._request_methods.RequestMethods
```
Comment (sethmlarson) [inline docs/reference/urllib3.request.rst:4]: `RequestMethods` is a private API, where are we publically mentioning it? Inheritance chains?

File: docs/reference/urllib3.response.rst [hunk 4]
```diff
@@ -4,10 +4,20 @@ Response and Decoders
...
+.. autoclass:: urllib3.response.BaseHTTPResponse
+    :members:
+    :undoc-members:
+    :show-inheritance:
+
 .. autoclass:: urllib3.response.HTTPResponse
     :members:
     :undoc-members:
     :show-inheritance:
+    :inherited-members:
```
Comment (sethmlarson) [inline docs/reference/urllib3.response.rst:16]: If possible we want to keep inherited members private except for common ones like `read()`. Is there a way we can opt-in to the members we want?

File: docs/reference/urllib3.util.rst [hunk 5]
```diff
@@ -4,14 +4,20 @@ Utilities
...
+At the very core, just like its predecessors, urllib3 is built on top of
 :mod:`http.client` -- the lowest level HTTP library included in the Python
 standard library.
 
-To aid the limited functionality of the :mod:`http.client` module, :mod:`urllib3`
+To aid the limited functionality of the :mod:`http.client` module, urllib3
 provides various helper methods which are used with the higher level components
 but can also be used independently.
 
+.. autoclass:: urllib3.util.timeout._TYPE_DEFAULT
```
Comment (sethmlarson) [inline docs/reference/urllib3.util.rst:15]: `_TYPE_*` types shouldn't be documented, they're private APIs.

**Review by sethmlarson — CHANGES_REQUESTED** [review]
Thanks for this!! Here's some thoughts:

File: docs/reference/urllib3.util.rst (hunk 5 above)
Comment (hramezani) [reply docs/reference/urllib3.util.rst:15]: Then we will have a lot of warnings like:

`WARNING: py:class reference target not found: urllib3.util.timeout._TYPE_DEFAULT`

File: docs/reference/urllib3.request.rst (hunk 3 above)
Comment (hramezani) [reply docs/reference/urllib3.request.rst:4]: Yes,  Inheritance chains of class `PoolManager`

File: docs/reference/urllib3.response.rst (hunk 4 above)
Comment (hramezani) [reply docs/reference/urllib3.response.rst]: Yes, we just have a warning for `json`. I've limited it to `json`

**Comment by hramezani** [comment]
@sethmlarson I've addressed your comments in a separate commit.
Regarding the remaining warnings, maybe we can consider @pquentin suggestion

> do we want to show type annotation at all in docs? If we don't then this will fix issues such as urllib3._collections.RecentlyUsedContainer not being documented.

**Comment by sethmlarson** [comment]
Definitely agree with Quentin's suggestion, private interfaces shouldn't be in the docs.

**Comment by hramezani** [comment]
- Added `nitpick_ignore` and ignore some private types
- Added `autodoc_typehints = "description"` as @pquentin suggested.

Here are the remaining warnings that I think can be ignored.

```
urllib3/exceptions.py:docstring of urllib3.exceptions.HeaderParsingError:: WARNING: py:class reference target not found: email.errors.MessageDefect
urllib3/response.py:docstring of urllib3.response.HTTPResponse:: WARNING: py:class reference target not found: http.client.HTTPMessage
/home/docs/.pyenv/versions/3.7.9/lib/python3.7/ssl.py:docstring of ssl.SSLContext:1: WARNING: py:class reference target not found: _ssl._SSLContext
urllib3/util/__init__.py:docstring of urllib3.util.ssl_.resolve_cert_reqs:: WARNING: py:class reference target not found: VerifyMode
urllib3/util/__init__.py:docstring of urllib3.util.ssl_.ssl_wrap_socket:: WARNING: py:class reference target not found: SSLTransportType

```

**Comment by pquentin** [comment]
`urllib3.request` does look much better:

Capture d’écran 2022-05-14 à 21 12 13

However we still have private APIs in type hints, they just don't have links:

Capture d’écran 2022-05-14 à 22 28 05

I think that's a good enough compromise if that's OK for @sethmlarson too. Because the alternative would be no type hints in docs at all I think?

**Comment by sethmlarson** [comment]
This looks great! :tada: Thanks for pushing this forward @hramezani and @pquentin

**Comment by hramezani** [comment]
I've updated the PR and added the remaining warnings to `nitpick_ignore`. but doc fails with the following error:

`pip._vendor.packaging.markers.UndefinedEnvironmentName: 'extra' does not exist in evaluation environment.`

**Review by sethmlarson — DISMISSED** [review]
Swapping my review status to approval so this can get merged as soon as docs build is fixed.

File: noxfile.py [hunk 6]
```diff
@@ -173,4 +173,4 @@ def docs(session: nox.Session) -> None:
     session.chdir("docs")
     if os.path.exists("_build"):
         shutil.rmtree("_build")
-    session.run("sphinx-build", "-b", "html", "-W", ".", "_build/html")
+    session.run("sphinx-build", "-b", "html", ".", "_build/html")
```
Comment (pquentin) [inline noxfile.py:176]: Can you please restore the `-W` flag?

File: src/urllib3/connectionpool.py [hunk 7]
```diff
@@ -558,7 +558,7 @@ def urlopen(  # type: ignore[override]
         .. note::
 
            More commonly, it's appropriate to use a convenience method provided
-           by :class:`.RequestMethods`, such as :meth:`request`.
+           by :class:`urllib3._request_methods.RequestMethods`, such as :meth:`request`.
```
Comment (pquentin) [inline src/urllib3/connectionpool.py:561]: Can you please remove "provided by urllib3._request_methods.RequestMethods"? It's a private API, and we already link to urllib3.request().

**Review by pquentin — CHANGES_REQUESTED** [review]
Thanks! We're nearly there

**Comment by hramezani** [comment]
Thanks @pquentin for the review. I've addressed your comments.

**Review by pquentin — APPROVED** [review]
Thank you! :tada:

**Pull request merged**